architecture.md

```diff
@@ -0,0 +1,91 @@
+For any other functionality, like sensor data processing for example, an additional system would have to be loaded.
+For example, if you need to generate sensor data that utilizes rendering sensors, you would need to load the sensors system.
+The visualization of that data, however, is left up to the client side plugins.
+
+## Communication process
+
+Any information that crosses the process boundary (whether back-to-front with scene broadcaster or front-to-back with user commands) has to go through [Ignition Transport](https://ignitionrobotics.org/libs/transport), the communication library.
+Synchronization between processes is performed by scene broadcaster, a server-side Ignition Gazebo plugin that uses Ignition Transport and [Ignition Messages](https://ignitionrobotics.org/libs/msgs) libraries to send messages from the server to the client.
+The messages themselves are provided by Ignition Messages, while the framework for creating the publishers and subscribers that exchange messages is from Ignition Transport.
+
+The scene broadcaster system is responsible for getting the state of the world (all of the entities and their component values) from the ever-changing simulation loop running in the back end, packaging that information into a very compact message, and periodically sending it to the frontend to process.
+Omitting the scene broadcaster, while possible, would mean not being able to visualize data on the frontend, which can be useful for saving computational power.
+
+In addition to state messages sent from the scene broadcaster to the client, the client can also make requests to the server using the user commands system (another backend plugin).
+It also utilizes the Ignition Transport and Ignition Messages libraries.
+The user command system has services that can be requested by the frontend GUI to insert, delete, move, etc., and will send responses back to the GUI acknowledging the receipt and execution of those requests.
+
+Several non-default plugins are able to send or receive messages to other processes outside of the server-client process.
+For example, the server-side sensors plugin publishes messages outside of the server-client process, and the server-side diff-drive plugin receives info from processes other than the client.
+One such process these plugins may communicate with, for example, is ROS.
+
+## Frontend client process
+
+The client-side process, essentially the GUI, also comprises multiple plugins, some loaded by default, others that can be added, and all optional.
+All of the frontend plugins use the [Ignition GUI](https://ignitionrobotics.org/libs/gui) library.
+The client itself also relies on Ignition GUI.
+There are visualization plugins that create the windows of the GUI, add buttons and other interactive features, and a 3D scene plugin that utilizes [Ignition Rendering](https://ignitionrobotics.org/libs/rendering) that creates the scene the user sees.
+
+Frontend plugins communicate using events.
```

Comment:
Can we define/describe what an "event" is?

Reply:
Good idea. Maybe something like

> Frontend plugins typically communicate among themselves using [events](https://ignitionrobotics.org/api/gui/4.2/namespaceignition_1_1gui_1_1events.html). Events are similar to messages, but they're processed synchronously. For example, the `Render` event is emitted by a 3D scene from it's rendering thread right before the scene is rendered; this gives other plugins the chance to execute code right at that thread at that moment, which is valuable to edit the 3D scene. Other such events are emitted when the user right-clicks or hovers the scene for example.